File: app/src/main/java/br/com/victoramaral/githubdive/view/activities/RepositoriesActivity.java
package br.com.victoramaral.githubdive.view.activities;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import br.com.victoramaral.githubdive.R;
import br.com.victoramaral.githubdive.model.pojos.repositories.Item;
import br.com.victoramaral.githubdive.view.adapter.RepositoryAdapter;
import br.com.victoramaral.githubdive.view.interfaces.RepositoryOnClick;
import br.com.victoramaral.githubdive.viewmodel.RepositoriesViewModel;

public class RepositoriesActivity extends AppCompatActivity implements RepositoryOnClick {

    private RecyclerView recyclerView;
    private RepositoriesViewModel viewModel;
    private RepositoryAdapter adapter;
    private List<Item> itemList = new ArrayList<>();
    private int page = 1;
    private int perPage = 10;

    public static final String CREATOR_KEY = "123456";
    public static final String REPOSITORY_KEY = "654321";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_repositories);
        initView();

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        viewModel.getAllRepositories(page, perPage);
        viewModel.getListaRepositories().observe(this, items -> {
            adapter.setUpdate(items);
        });

        scrollListener();
    }

    public void initView() {
        recyclerView = findViewById(R.id.recyclerViewRepo);
        viewModel = ViewModelProviders.of(this).get(RepositoriesViewModel.class);
        adapter = new RepositoryAdapter(itemList, this);
    }

    public void scrollListener() {
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int
                    newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                LinearLayoutManager layoutManager = (LinearLayoutManager)
                        recyclerView.getLayoutManager();
                int totalItemCount = layoutManager.getItemCount();
                int lastVisible = layoutManager.findLastVisibleItemPosition();

                boolean endHasBeenReached = lastVisible + 5 >= totalItemCount;

                if (totalItemCount > 0 && endHasBeenReached) {
                    page++;
                    viewModel.getAllRepositories(page, perPage);
                }
            }
        });
    }

    @Override
    public void repositoryOnClick(Item item) {
        Toast.makeText(this, "Você selecionou o repositório " + item.getName(),
                Toast.LENGTH_LONG).show();
        Intent intent = new Intent(RepositoriesActivity.this,
                PullRequestActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString(CREATOR_KEY, item.getOwner().getLogin());
        bundle.putString(REPOSITORY_KEY, item.getName());
        intent.putExtras(bundle);
        startActivity(intent);
    }
}

File: GitRepositories/app/src/main/java/com/example/gitrepositories/modules/repositories/RepositoriesActivity.kt
package com.example.gitrepositories.modules.repositories

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.gitrepositories.R
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.repositories.*

class RepositoriesActivity : AppCompatActivity() {

    lateinit var viewModel: RepositoriesViewModel
    lateinit var adapter: RepositoriesAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.repositories)

        viewModel = ViewModelProviders.of(this).get(RepositoriesViewModel::class.java)
        setUpAdapter()
        setUpObservers()
    }

    private fun setUpAdapter() {
        adapter = RepositoriesAdapter(this, viewModel::onRepositoryClick)
        repositories_list.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        repositories_list.adapter = adapter
    }

    private fun setUpObservers() {
        viewModel.list.observe(this, Observer {
            adapter.submitList(it)
        })

        viewModel.intent.observe(this, Observer {
            startActivity(it)
        })

        viewModel.displayEmptyMessage.observe(this, Observer {
            empty_list_text.visibility = if (it) View.VISIBLE else View.GONE
        })

        viewModel.displayConnectivityMessage.observe(this, Observer {
            Snackbar.make(repositories_layout, it, Snackbar.LENGTH_LONG).show()
        })
    }
}